proposals/nnnn-inferring-senable-for-methods

```diff
@@ -0,0 +1,212 @@
+# [Pitch] Inferring `@Sendable` for methods
+
+* Proposal: [SE-NNNN](https://github.com/kavon/swift-evolution/blob/sendable-functions/proposals/NNNN-filename.md)
+* Authors: [Angela Laar](https://github.com/angela-laar), [Kavon Farvardin](https://github.com/kavon)
+* Review Manager: TBD
+* Status: Awaiting Implementation
+
+## Introduction
+
+This proposal is focused on a few corner cases in the language surrounding functions as values when using concurrency. The goal is to improve flexibility, simplicity, and ergonomics without significant changes to Swift.
+
+## Motivation
+
+The partial application of methods and other first-class uses of functions have a few rough edges when combined with concurrency.
+
+Let’s look at partial application on its own before we combine it with concurrency.  In Swift, you can create a function-value representing a method by writing an expression that only accesses (but does not call) a method using one of its instances. This access is referred to as a "partial application" of a method to one of its (curried) arguments - the object instance.
+
+```
+struct S {
+  func f() { ... }
+}
+
+let partial: (() -> Void) = S().f 
+```
+
+
+When referencing a method *without* partially applying it  to the object instance, using the expression NominalType.method, we call it "unapplied."
+
+
+```
+let unapplied:(T) -> (() -> Void) = S.f 
+```
+
+
+Suppose we want to create a generic method that expects an unapplied function method conforming to Senable as a parameter. We can create a protocol ``P`` that conforms to the `Sendable` protocol and tell our generic function to expect some generic type that conforms to ``P``. We can also use the `@Sendable` attribute, introduced for closures and functions in [SE-302](https://github.com/kavon/swift-evolution/blob/sendable-functions/proposals/0302-concurrent-value-and-concurrent-closures.md), to annotate the closure parameter. 
+
+
+```
+protocol P: Sendable {
+  init()
+}
+
+func g<T>(_ f: @escaping @Sendable (T) -> (() -> Void)) where T: P {
+  Task {
+    let instance = T()
+    f(instance)()
+  }
+}
+```
+
+Now let’s call our method and pass our struct type `S` . First we should make `S` conform to Sendable, which we can do by making `S` conform to our new Sendable type `P` . 
+
+This should make `S` and its methods Sendable as well. However, when we pass our unapplied function `S.f`  to our generic function `g`, we get a warning that `S.f` is not Sendable as `g()` is expecting.  
+
+
+```
+struct S: P {
+  func f() { ... }
+}
+
+g(S.f) // Converting non-sendable function value to '@Sendable (S) -> (() -> Void)' may introduce data races
+```
+
+
+We can work around this by wrapping our unapplied function in a Sendable closure.  
+
+```
+// S.f($0) == S.f()
+g({ @Sendable **in** S.f($0) })
+```
+
+
+This is a lot of churn to get the expected behavior. The compiler should preserve `@Sendable` in the type signature instead.
+
+## Proposed solution
```

Comment:
I would start this section with a straightforward description of what you want to change:

```suggestion
## Proposed solution

We propose that the compiler should automatically infer `@Sendable` on functions which cannot capture non-`Sendable` state. These include:

* Partially-applied and unapplied instance methods of `Sendable` types
* Non-local functions

The compiler should furthermore ban use of `@Sendable` on instance methods of non-`Sendable` types.
```

The explanation that follows is useful as justification for that change, but it's best to state up front what it is that you're justifying.